### PR TITLE
EES-119: Create methodology

### PR DESCRIPTION
**Pull request by jamescutts**
Adds new page for creating methodology.

Adds new dashboard page for listing methodology

### CODE REVIEW

File: src/explore-education-statistics-admin/src/pages/admin-dashboard/AdminDashboardPage.tsx [hunk 1]
```diff
@@ -92,7 +92,19 @@ const AdminDashboardPage = ({ match }: RouteComponentProps<MatchProps>) => {
                   </li>
                 </span>
               </h1>
+
+              <p>This is your administration dashboard here you can:</p>
```
Comment (duncan-at-hiveit) [inline src/explore-education-statistics-admin/src/pages/admin-dashboard/AdminDashboardPage.tsx]: Should this have a hyphen in it?

`This is your administration dashboard - here you can:`

File: src/explore-education-statistics-admin/src/pages/methodology/createMethodologyPage.tsx [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+}
+
+const CreateMethodologyPage = ({
+  match,
+  history,
+}: RouteComponentProps<MatchProps>) => {
+  const [model, setModel] = useState<CreateMethodologyModel>();
+
+  useEffect(() => {
+    Promise.all([service.getPublicationAndReleaseContacts()]).then(
```
Comment (duncan-at-hiveit) [inline src/explore-education-statistics-admin/src/pages/methodology/createMethodologyPage.tsx]: Only one promise occurring here, so don't need the Promise.all().

Can just write:

`service.getPublicationAndReleaseContacts().then(contacts => `

File: src/explore-education-statistics-admin/src/pages/methodology/createMethodologyPage.tsx [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+      title: values.methodologyTitle,
+      publishScheduled: dayMonthYearInputsToDate(values.scheduledPublishDate),
+      contactId: values.selectedContactId,
+    };
+
+    const createdMethodology = await methodologyService.createMethodology(
+      submission,
+    );
+
+    // TODO: redirect to the methodology summary
```
Comment (duncan-at-hiveit) [inline src/explore-education-statistics-admin/src/pages/methodology/createMethodologyPage.tsx]: Is this TODO not TODOne?

File: src/explore-education-statistics-admin/src/pages/methodology/createMethodologyPage.tsx [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+    >
+      <div className="govuk-grid-row">
+        <div className="govuk-grid-column-two-thirds">
+          <h1 className="govuk-heading-xl">Create new methodology</h1>
+        </div>
+        <div className="govuk-grid-column-one-third">
+          <RelatedInformation heading="Help and guidance">
+            <ul className="govuk-list">
+              <li>
+                <Link to="/documentation" target="blank">
```
Comment (duncan-at-hiveit) [inline src/explore-education-statistics-admin/src/pages/methodology/createMethodologyPage.tsx]: Does this need a specific URL?

Comment (duncan-at-hiveit) [reply src/explore-education-statistics-admin/src/pages/methodology/createMethodologyPage.tsx]: And does "blank" need to be "_blank"?

File: src/explore-education-statistics-admin/src/pages/methodology/createMethodologyPage.tsx [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+      {model && (
+        <Formik<FormValues>
+          enableReinitialize
+          initialValues={{
+            methodologyTitle: '',
+            selectedContactId: orderBy(
+              model.contacts,
+              contact => contact.contactName,
+            )[0].id,
+            scheduledPublishDate: dayMonthYearValuesToInputs(
```
Comment (duncan-at-hiveit) [inline src/explore-education-statistics-admin/src/pages/methodology/createMethodologyPage.tsx]: Are we doing anything with schedulePublished, cos most of it is commented out in the backend

File: src/explore-education-statistics-admin/src/pages/methodology/editMethodologyPage.tsx [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+            [[methodology title]]
+          </h1>
+        </div>
+
+        <div className="govuk-grid-column-one-third">
+          <RelatedInformation heading="Help and guidance">
+            <ul className="govuk-list">
+              <li>
+                <Link to="/documentation" target="blank">
+                  Creating new methodology{' '}
```
Comment (duncan-at-hiveit) [inline src/explore-education-statistics-admin/src/pages/methodology/editMethodologyPage.tsx]: This is for editing an existing Methodology, so is this documentation relevant?

File: src/explore-education-statistics-admin/src/pages/prototypes/PrototypeAdminDashboard.tsx [hunk 7]
```diff
@@ -78,45 +87,221 @@ const PrototypeBrowseReleasesPage = ({ location }: RouteChildrenProps) => {
           </RelatedInformation>
         </div>
       </div>
-      <Tabs id="dashboard-tabs">
-        <TabsSection
-          id="publications"
-          title={
-            userContext.user &&
+
+      <p>This is your administration dashboard here you can:</p>
```
Comment (duncan-at-hiveit) [inline src/explore-education-statistics-admin/src/pages/prototypes/PrototypeAdminDashboard.tsx]: Hyphen

File: src/explore-education-statistics-admin/src/pages/methodology/createMethodologyPage.tsx [hunk 8]
```diff
@@ -0,0 +1,223 @@
+import Link from '@admin/components/Link';
```
Comment (duncan-at-hiveit) [inline src/explore-education-statistics-admin/src/pages/methodology/createMethodologyPage.tsx]: Pascal-case the file names that export Components (i.e. createMethodolodyPage.tsx => CreateMethodologyPage.tsx) in keeping with the other Components

File: src/explore-education-statistics-admin/src/services/methodology/types.ts [hunk 9]
```diff
@@ -9,3 +9,9 @@ export interface MethodologyStatusPublication {
   id: string;
   title: string;
 }
+
+export interface CreateMethodologyRequest {
+  title: string;
+  publishScheduled: Date;
```
Comment (duncan-at-hiveit) [inline src/explore-education-statistics-admin/src/services/methodology/types.ts]: We still using this, as most usages in the backend are commented out

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Services/MethodologyService.cs [hunk 10]
```diff
@@ -22,24 +25,51 @@ public MethodologyService(ContentDbContext context, IMapper mapper)
...
+            CreateMethodologyViewModel methodology)
+        {
+            return await ValidateMethodologySlugUnique(methodology.Slug)
+                .OnSuccess(async () =>
+                {
+            var model = new Methodology
+            {
+                Title = methodology.Title,
+                Slug = methodology.Slug,
+                // PublishScheduled = methodology.PublishScheduled
```
[the remark this answers is not in the document]
Comment (jamescutts) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/Services/MethodologyService.cs]: Its not currently stored, though it should be, i was holding off any DB change into the current PRs where merged

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Controllers/Api/MethodologiesController.cs [hunk 11]
```diff
@@ -20,18 +21,44 @@ public MethodologiesController(IMethodologyService methodologyService)
             _methodologyService = methodologyService;
         }
 
-        // GET api/topic/{topicId}/methodologies
+        [Produces("application/json")]
```
[the remark this answers is not in the document]
Comment (jamescutts) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/Controllers/Api/MethodologiesController.cs]: Interesting, if shall have a look and see

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Services/MethodologyService.cs [hunk 12]
```diff
@@ -49,5 +79,15 @@ public async Task<List<MethodologyViewModel>> GetTopicMethodologiesAsync(TopicId
                 .ToListAsync();
             return _mapper.Map<List<MethodologyViewModel>>(methodologies);
         }
+        
+        private async Task<Either<ValidationResult, bool>> ValidateMethodologySlugUnique(string slug)
+        {
+            if (await _context.Methodologies.AnyAsync(r => r.Slug == slug))
+            {
+                return new ValidationResult(ValidationErrorMessages.SlugNotUnique.ToString());
```
[the remark this answers is not in the document]
Comment (jamescutts) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/Services/MethodologyService.cs]: Yeah i think that was the reason

File: src/explore-education-statistics-admin/src/pages/methodology/createMethodologyPage.tsx (hunk 5 above)
Comment (jamescutts) [reply src/explore-education-statistics-admin/src/pages/methodology/createMethodologyPage.tsx]: See comment above :)

File: src/explore-education-statistics-admin/src/pages/methodology/editMethodologyPage.tsx (hunk 6 above)
Comment (jamescutts) [reply src/explore-education-statistics-admin/src/pages/methodology/editMethodologyPage.tsx]: We dont currently have any methodology documentation so directing to generic documentation page. however link naming is consitient with release.

File: src/explore-education-statistics-admin/src/services/methodology/types.ts (hunk 9 above)
Comment (jamescutts) [reply src/explore-education-statistics-admin/src/services/methodology/types.ts]: As above comments :)

File: src/explore-education-statistics-admin/src/pages/methodology/createMethodologyPage.tsx (hunk 4 above)
Comment (jamescutts) [reply src/explore-education-statistics-admin/src/pages/methodology/createMethodologyPage.tsx]: Specific URL does not currently exist so pointing to general documentation for now

**Comment by jamescutts** [comment]
Think this is missing the migration for the PublishScheduled Field on Methodology

**Pull request merged**